### PR TITLE
Adds `ShaderStorageBuffer` asset

### PR DESCRIPTION
**Pull request by tychedelia**
Adds a new `Handle<Storage>` asset type that can be used as a render asset, particularly for use with `AsBindGroup`.

Closes: #13658 

# Objective

Allow users to create storage buffers in the main world without having to access the `RenderDevice`. While this resource is technically available, it's bad form to use in the main world and requires mixing rendering details with main world code. Additionally, this makes storage buffers easier to use with `AsBindGroup`, particularly in the following scenarios:
- Sharing the same buffers between a compute stage and material shader. We already have examples of this for storage textures (see game of life example) and these changes allow a similar pattern to be used with storage buffers.
- Preventing repeated gpu upload (see the previous easier to use `Vec` `AsBindGroup` option).
- Allow initializing custom materials using `Default`. Previously, the lack of a `Default` implement for the raw `wgpu::Buffer` type made implementing a `AsBindGroup + Default` bound difficult in the presence of buffers.

## Solution

Adds a new `Handle<Storage>` asset type that is prepared into a `GpuStorageBuffer` render asset. This asset can either be initialized with a `Vec<u8>` of properly aligned data or with a size hint. Users can modify the underlying `wgpu::BufferDescriptor` to provide additional usage flags.

## Migration Guide

The `AsBindGroup` `storage` attribute has been modified to reference the new `Handle<Storage>` asset instead. Usages of Vec` should be converted into assets instead.

### CODE REVIEW

File: examples/shader/storage.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    // Camera
+    commands.spawn(Camera3dBundle {
+        transform: Transform::from_xyz(0.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
+        ..default()
+    });
+}
+
+// Update the material color by time
+fn update(
```
Comment (IceSentry) [inline examples/shader/storage.rs:67]: This should probably show how to update the `Handle<Storage>`

Comment (tychedelia) [reply examples/shader/storage.rs:66]: Added a better example that focuses more on the buffer.

**Comment by JMS55** [comment]
See related old PR: #6669

File: crates/bevy_internal/src/default_plugins.rs [hunk 2]
```diff
@@ -30,6 +30,8 @@ plugin_group! {
         #[cfg(feature = "bevy_render")]
         bevy_render::texture:::ImagePlugin,
         #[cfg(feature = "bevy_render")]
+        bevy_render::storage:::StoragePlugin,
```
Comment (JMS55) [inline crates/bevy_internal/src/default_plugins.rs:33]: Why is this here, and not in the top-level plugin for bevy_render?

Also if you keep it here, there's a list of default plugins somewhere that needs updating to match.

Comment (tychedelia) [reply crates/bevy_internal/src/default_plugins.rs:33]: This is the list of default plugins, unless you mean something else. I added it here because it's next to the equivalent image asset plugin.

Comment (JMS55) [reply crates/bevy_internal/src/default_plugins.rs:33]: Usually we have one plugin per bevy crate that wraps all other plugins in that crate and is used for the default plugin set. I.e. bevy_pbr::PbrPlugin. Not sure why bevy_render is split up though.

Comment (tychedelia) [reply crates/bevy_internal/src/default_plugins.rs:33]: Ah, it makes sense to me now, it's because it allows easy overriding the `ImagePlugin` at the top level. This isn't the case for this new plugin, so I'll move it into `RenderPlugin`. 👍

File: examples/shader/storage.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    time: Res<Time>,
+    material_handle: Res<CustomMaterialHandle>,
+    mut materials: ResMut<Assets<CustomMaterial>>,
+    mut buffers: ResMut<Assets<Storage>>,
+) {
+    let material = materials.get_mut(&material_handle.0).unwrap();
+    material.colors = buffers.add(Storage::new(
```
Comment (IceSentry) [inline examples/shader/storage.rs:74]: is it possible to update the existing buffer instead of creating a new one?

File: assets/shaders/storage.wgsl [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+@vertex
+fn vertex(vertex: Vertex) -> VertexOutput {
+    var out: VertexOutput;
+    var world_from_local = mesh_functions::get_world_from_local(vertex.instance_index);
+    out.world_position = mesh_functions::mesh_position_local_to_world(world_from_local, vec4(vertex.position, 1.0));
+    out.clip_position = position_world_to_clip(out.world_position.xyz);
+
+    out.color = colors[vertex.instance_index % 5];
```
Comment (IceSentry) [inline assets/shaders/storage.wgsl:26]: Maybe add a quick comment on why this is necessary and what this does?

File: crates/bevy_render/src/render_resource/bind_group.rs [hunk 5]
```diff
@@ -296,9 +297,16 @@ pub trait AsBindGroup {
         render_device: &RenderDevice,
         images: &RenderAssets<GpuImage>,
         fallback_image: &FallbackImage,
+        buffers: &RenderAssets<GpuStorageBuffer>,
```
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/bind_group.rs:300]: I was already annoyed that this function signature was too long, but I guess it's not too bad. I'd kinda like if it was optional though. I've had a bunch of situations where I tried using AsBindGroup without an image/texture and it was annoying to have to provide it image stuff.

Comment (IceSentry) [reply crates/bevy_render/src/render_resource/bind_group.rs:300]: I guess this could be a potential footgun, but I feel people being this deep inside bevy's rendering api and using it directly would know what they are doing enought that it matters a bit less.

**Comment by IceSentry** [comment]
Like mentioned on discord. I'd prefer if this kept the ability to use a raw `Buffer` with `AsBindGroup`.

File: examples/shader/storage_buffer.rs [hunk 6]
```diff
@@ -0,0 +1,113 @@
+//! This example demonstrates how to use a storage buffer with `AsBindGroup` in a custom material.
```
Comment (IceSentry) [inline examples/shader/storage_buffer.rs:1]: Should the example be renamed to match the new type? I don't really have an opinion either way honestly.

Comment (tychedelia) [reply examples/shader/storage_buffer.rs:1]: My thought was just that `shader/shader_storage_buffer` was a bit redundant.

Comment (tychedelia) [reply examples/shader/storage_buffer.rs:1]: But happy to change :)

File: crates/bevy_render/src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    fn default() -> Self {
+        Self {
+            data: None,
+            buffer_description: wgpu::BufferDescriptor {
+                label: None,
+                size: 0,
+                usage: BufferUsages::STORAGE,
+                mapped_at_creation: false,
+            },
+            asset_usage: RenderAssetUsages::empty(),
```
Comment (IceSentry) [inline crates/bevy_render/src/storage.rs:48]: Why empty? Shouldn't this be all()?

**Review by IceSentry — APPROVED** [review]
I didn't test the example, but the implementation LGTM

**Comment by IceSentry** [comment]
Also, could you update the description with the new name?

File: crates/bevy_render/src/storage.rs (hunk 7 above)
Comment (tychedelia) [reply crates/bevy_render/src/storage.rs]: It's wrong, should be default.

File: Cargo.toml [hunk 8]
```diff
@@ -2415,6 +2415,17 @@ description = "A shader that shows how to bind and sample multiple textures as a
 category = "Shaders"
 wasm = false
 
+[[example]]
+name = "storage"
```
Comment (IceSentry) [inline Cargo.toml:2419]:
```suggestion
name = "storage_buffer"
```

File: examples/shader/storage_buffer.rs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                ..default()
+            });
+        }
+    }
+
+    // Camera
+    commands.spawn(Camera3dBundle {
+        transform: Transform::from_xyz(0.0, 0.0, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
+        ..default()
+    });
```
Comment (IceSentry) [inline examples/shader/storage_buffer.rs:63]: Small suggestion to make things a bit nicer to look at

```suggestion
    for i in -6..=6 {
        for j in -3..=3 {
            commands.spawn(MaterialMeshBundle {
                mesh: meshes.add(Cuboid::from_size(Vec3::splat(0.3))),
                material: material_handle.clone(),
                transform: Transform::from_xyz(i as f32, j as f32, 0.0),
                ..default()
            });
        }
    }

    // Camera
    commands.spawn(Camera3dBundle {
        transform: Transform::from_xyz(0.0, 0.0, 10.0).looking_at(Vec3::ZERO, Vec3::Y),
        ..default()
    });
```

**Review by ChristopherBiscardi — APPROVED** [review]
ran the example successfully and the new usage makes sense to me
